kinto/views/buckets.py

```diff
@@ -1,8 +1,70 @@
+from pyramid.httpexceptions import HTTPForbidden, HTTPPreconditionFailed
+from pyramid.security import NO_PERMISSION_REQUIRED
+from pyramid.view import view_config
+
 from cliquet import resource
+from cliquet.utils import hmac_digest
+from cliquet.views.batch import build_request
 
 from kinto.views import NameGenerator
 
 
+@view_config(route_name='default_bucket', permission=NO_PERMISSION_REQUIRED)
+def default_bucket(request):
+    if getattr(request, 'prefixed_userid', None) is None:
+        raise HTTPForbidden  # Pass through the forbidden_view_config
+
+    settings = request.registry.settings
+    hmac_secret = settings['cliquet.userid_hmac_secret']
+    # Build the user unguessable bucket_id UUID from its user_id
+    bucket_id = hmac_digest(hmac_secret, request.prefixed_userid)[:32]
+    path = request.path.replace('default', bucket_id)
+
+    # Make sure bucket exists
+    is_bucket_put = (request.method.lower() == 'put' and
+                     request.path.endswith('buckets/default'))
+    if not is_bucket_put:
+        subrequest = build_request(request, {
+            'method': 'PUT',
+            'path': '/buckets/%s' % bucket_id,
+            'body': {"data": {}},
+            'headers': {'If-None-Match': '*'}
+        })
+        try:
+            request.invoke_subrequest(subrequest)
+        except HTTPPreconditionFailed:
+            # The bucket already existed
+            pass
+
+    # Make sure the collection exists
+    subpath = request.matchdict['subpath']
+    if subpath.startswith('/collections/'):
+        collection_id = subpath.split('/')[2]
+        is_collection_put = (request.method.lower() == 'put' and
+                             request.path.endswith(collection_id))
+        if not is_collection_put:
+            subrequest = build_request(request, {
+                'method': 'PUT',
+                'path': '/buckets/%s/collections/%s' % (
+                    bucket_id, collection_id),
+                'body': {"data": {}},
+                'headers': {'If-None-Match': '*'}
+            })
+            try:
+                request.invoke_subrequest(subrequest)
+            except HTTPPreconditionFailed:
+                # The collection already existed
+                pass
```

Comment:
We could put the bucket check+creation and collection check+creation in separate functions to ease readability and have a smaller function.